simtools/util/workflow_description.py

```diff
@@ -0,0 +1,571 @@
+                _filename += '-' + self.workflow_config['ACTIVITY']['NAME']
+        except KeyError:
+            self._logger.error(
+                "Missing CTA:PRODUCT:ID in metadata")
+            raise
+
+        if not suffix:
+            suffix = '.' + self.product_data_file_format(suffix=True)
+
+        return Path(_directory).joinpath(_filename+suffix)
+
+    def product_data_file_format(self, suffix=False):
+        """
+        Return file format for data file
+
+        Parameter
+        ---------
+        suffix: bool
+            return file suffix (if true)
+            return file format (if false)
+
+        Returns
+        -------
+        str
+            file format of data product; default file format is 'ascii.ecsv'
+
+        Raises
+        ------
+        KeyError
+            if relevant fields are not defined in toplevel metadata
+            dictionary
+
+        """
+
+        _file_format = "ascii.ecsv"
+        try:
+            _file_format = self.workflow_config['PRODUCT']['FORMAT']
+        except KeyError:
+            self._logger.info(
+                "Using default file format for model file: ascii.ecsv")
+
+        if suffix and _file_format == "ascii.ecsv":
+            _file_format = "ecsv"
+
+        return _file_format
+
+    def product_data_directory(self):
+        """
+        Return output directory for data products.
+        Create directory if necessary.
+
+        Output directory is determined following this sorted
+        list:
+        1. PRODUCT:DIRECTORY is set (e.g., through workflow file)
+        2. gammasim-tools output location
+
+        Returns
+        -------
+        str
+            output directory for data products
+
+        Raises
+        ------
+        KeyError
+            if data file name is not defined in workflow configuration
+            or in product metadata dict
+
+        """
+
+        _output_label = self.workflow_config['ACTIVITY']['NAME']
+
+        if self.workflow_config['PRODUCT']['DIRECTORY']:
+            path = Path(self.workflow_config['PRODUCT']['DIRECTORY'])
+            path.mkdir(parents=True, exist_ok=True)
+            _output_dir = path.absolute()
+        else:
+            _output_dir = cfg.get("outputLocation")
+
+        _output_dir = io.getApplicationOutputDirectory(
+            _output_dir, _output_label)
+
+        self._logger.info("Outputdirectory {}".format(_output_dir))
+        return _output_dir
+
+    def _from_args(self, key, default_return=None):
+        """
+        Return argparser argument.
+        No errors raised if argument does not exist
+
+        """
+
+        try:
+            return self.args.__dict__[key]
+        except KeyError:
+            pass
+
+        return default_return
+
+    def _fill_toplevel_meta_from_args(self):
+        """
+        Fill metadata available through command line into top-level template
+
+        Parameters
+        ----------
+        args: argparse.Namespace
```

Comment:
That is not a parameter you pass to the function. Either we should pass it, or if it is only relevant to the current instance of the class, remove this.

Reply:
Removed. 5d898e7